### PR TITLE
[qwen-vl] fix position ids

### LINKED ISSUE #40136: Qwen2.5-VL-7B-Instruct: Significant accuracy regression on MMMU benchmark with transformers >=4.54.0
## Description

We've observed a significant accuracy regression when evaluating Qwen2.5-VL-7B-Instruct on the MMMU Literature benchmark after upgrading from transformers 4.53.3 to 4.54.0. The accuracy drops from **86.67%** to **73.33%**, representing a **13.34 percentage point decrease** (~15.4% relative regression).

## Environment

- **Model**: Qwen/Qwen2.5-VL-7B-Instruct
- **Evaluation framework**: lm-eval
- **Task**: mmmu_val_literature (30 samples)
- **Hardware**: CUDA-enabled GPU
- **Python**: 3.10
- **Other dependencies**: torch, torchvision, accelerate, Pillow

### Transformers 4.53.3 Output
```
hf-multimodal (pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True), gen_kwargs: (None), limit: None, num_fewshot: 0, batch_size: 8
|  Tasks   |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|----------|------:|------|-----:|------|---|-----:|---|-----:|
|Literature|      0|none  |     0|acc   |↑  |0.8667|±  |0.0631|
```

### Transformers 4.54.0 Output
```
hf-multimodal (pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True), gen_kwargs: (None), limit: None, num_fewshot: 0, batch_size: 8
|  Tasks   |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|----------|------:|------|-----:|------|---|-----:|---|-----:|
|Literature|      0|none  |     0|acc   |↑  |0.7333|±  |0.0821|
```

## Summary

| Version | Accuracy | Stderr | Difference from baseline |
|---------|----------|--------|-------------------------|
| 4.53.3  | 0.8667   | ±0.0631| -                       |
| 4.54.0  | 0.7333   | ±0.0821| -0.1334 (-15.4%)        |

## Possible causes

The regression may be related to:
- Changes in the default image processor (slow → fast) 
- Changes in the vision model loading mechanism
- Modifications to the multimodal pipeline

## Impact

This regression affects production systems using Qwen2.5-VL models for visual question answering and multimodal understanding tasks. Users upgrading to 4.54.0 may experience unexpected accuracy drops.

## Additional context

- The evaluation uses a fixed random seed for reproducibility
- Both tests were run on the same hardware with identical configurations
- The regression is statistically significant given the standard errors

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### 1. Install dependencies
```bash
pip install lm-eval torch torchvision accelerate Pillow
```

### 2. Test with transformers 4.53.3 (baseline)
```bash
pip install transformers==4.53.3

lm_eval \
    --model hf-multimodal \
    --model_args "pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True" \
    --tasks mmmu_val_literature \
    --num_fewshot 0 \
    --batch_size 8 \
    --verbosity INFO
```

### 3. Test with transformers 4.54.0
```bash
pip install transformers==4.54.0

lm_eval \
    --model hf-multimodal \
    --model_args "pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True" \
    --tasks mmmu_val_literature \
    --num_fewshot 0 \
    --batch_size 8 \
    --verbosity INFO
```

## Expected behavior

The model accuracy should remain consistent or improve between minor version updates. A 15% relative performance regression on a standard benchmark is unexpected and concerning.

### LINKED ISSUE #40154: Qwen2.5VL is broken!
### System Info

- `transformers` version: 4.55.0
- Platform: Linux-5.15.0-1032-oracle-x86_64-with-glibc2.35
- Python version: 3.12.3
- Huggingface_hub version: 0.34.3
- Safetensors version: 0.6.1
- Accelerate version: 1.10.0
- Accelerate config:    not found
- DeepSpeed version: 0.17.4
- PyTorch version (accelerator?): 2.7.1+cu128 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

_No response_

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

In recent PR: https://github.com/huggingface/transformers/pull/39447

A text position embedding is introduced:

https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1624-R1628

```python
        if position_ids.ndim == 3 and position_ids.shape[0] == 4:
            text_position_ids = position_ids[0]
            position_ids = position_ids[1:]
        else:
            text_position_ids = position_ids[0]
```

The `text_position_ids` will be fed to the decoder layer: https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1665

```python
            layer_outputs = decoder_layer(
                hidden_states,
                attention_mask=causal_mask_mapping[decoder_layer.attention_type],
                position_ids=text_position_ids,
                past_key_value=past_key_values,
                output_attentions=output_attentions,
                use_cache=use_cache,
```

OK. So far we have learned:

1. If `position_ids` inputs as a 3-dim tensor with shape (4, B, seq_len), then `text_position_ids` is the first (1, B, seq_len).
2. Otherwise, it just copy the first `position_ids[0]`

Now, you can easily verify that this value in training time is a 3-dim tensor with shape (3, B, seq_len), as shown in this figure: (note that I have visual input so seq_len != max(pos_id)

Image

Therefore, text_position_ids is simply (0, ..., 631) as you would expect in previous version.

However, in generation time, the text_position_ids is prepared by:

https://github.com/huggingface/transformers/pull/39447/files#diff-41bc5f79048d22eb30688e5440a89f4303ed02aa3d7c99aa807e3e7243e1fc0aR861-R864

```python
            if "position_ids" not in model_inputs:
                text_positions = torch.arange(input_ids, device=input_ids.device)[None, None, :]
            else:
                text_positions = model_inputs["position_ids"][None, ...]
```

If we take a look ahead, model_inputs["position_ids"] is prepared by GenerationMixin. In both case, text_positions is default to the (0, ..., seq_len-1). Then, this text_positions (which is already WRONG because it doesn't take care of the offset caused by potential visual tokens) will be concat to form a 4D position_ids.

```python
model_inputs["position_ids"] = torch.cat([text_positions, vision_positions], dim=0)
```

Then, such a value will be sent to decoder layer.

In summary, in training time, the decoder layer will see position_ids to be offsetted by visual token. But in generation, it see a plain (0,...,seq_len-1) position_ids.

This might be the root cause to https://github.com/huggingface/transformers/issues/40136

### Expected behavior

I expect it works normally.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

 Fixes https://github.com/huggingface/transformers/issues/40136 and fixes https://github.com/huggingface/transformers/issues/40154

The accuracy on lm eval is restored. The issue was in position ids which weren't prepared correctly when generating. Using packed or unpacked positions doesn't affect anything, as I thought at first

Ig the values in position ids do not affect much generation results, given that slow tests weren't failing. We could add a test with the specific image/prompt where the outputs are significantly different, but I don't think it is necessary

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_vl, qwen2_5_vl

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_vl, qwen2_vl

**Comment by zucchini-nlp** [comment]
prob cc @vasqu for FA2 and packed attention

**Comment by zucchini-nlp** [comment]
run-slow: qwen2_5_omni, qwen2_5_vl, qwen2_vl

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py [hunk 1]
```diff
@@ -1586,7 +1586,7 @@ def forward(
             text_position_ids = position_ids[0]
             position_ids = position_ids[1:]
         else:
-            text_position_ids = position_ids[0]
+            text_position_ids = None
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py:1590]: Just fmi interest, was this some 2d tensor then? Or what was passed?

Would be nice to add a comment above, e.g. otherwise we ignore the text positions or something

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 2]
```diff
@@ -1308,7 +1308,7 @@ def forward(
                 else:
                     delta = torch.zeros((batch_size, seq_length), device=inputs_embeds.device)
                 delta = delta.repeat_interleave(batch_size // delta.shape[0], dim=1)
-                position_ids += delta.to(position_ids.device)
+                position_ids = position_ids + delta.to(position_ids.device)
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1312]: Hmm, any particular reason for inplace to outplace (no need to change)

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py [hunk 3]
```diff
@@ -1573,17 +1573,16 @@ def prepare_inputs_for_generation(
                 self.model.rope_deltas = rope_deltas
             # then use the prev pre-calculated rope-deltas to get the correct position ids
             elif "position_ids" in model_inputs:
-                position_ids = model_inputs["position_ids"][None, ...]
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1576]: So previously we already somewhat included a shift and now we leave the shifts all to the rope delta which shouldve been done from the start

File: src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py [hunk 4]
```diff
@@ -818,17 +818,16 @@ def prepare_inputs_for_generation(
...
+                delta = delta.repeat_interleave(batch_size // delta.shape[0], dim=0)
                 vision_positions = position_ids + delta.expand_as(position_ids)
-                vision_positions = vision_positions.expand(3, vision_positions.shape[1], -1)
 
             # Concatenate "text + vision" positions into [4, bs, seq-len]
-            if "position_ids" not in model_inputs:
-                text_positions = torch.arange(input_ids, device=input_ids.device)[None, None, :]
-            else:
-                text_positions = model_inputs["position_ids"][None, ...]
+            text_positions = model_inputs["position_ids"][None, ...]
```
Comment (vasqu) [inline src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py:830]: Nit: Eerily similar, was there no way to apply modular for this section on the 2.5 VL?

File: tests/test_modeling_common.py [hunk 5]
```diff
@@ -4767,6 +4767,15 @@ def update_config_headdim(config, requested_dim):
...
                 config.hidden_size *= max(requested_dim // head_dim, 1)
 
+                # Some models use 3D RoPE where the sum of RoPE sections has to be equal to head dim
+                if (
+                    getattr(config, "rope_scaling", None) is not None
+                    and config.rope_scaling.get("mrope_section") is not None
+                ):
+                    mrope_section = config.rope_scaling["mrope_section"]
+                    mutiplier = max(requested_dim // head_dim, 1)
+                    config.rope_scaling = {"type": "default", "mrope_section": [i * mutiplier for i in mrope_section]}
```
Comment (vasqu) [inline tests/test_modeling_common.py:4777]: Oh, similar idea we had :D hehe

**Review by vasqu — APPROVED** [review]
Just nits and question for my understanding

> We could add a test with the specific image/prompt where the outputs are significantly different, but I don't think it is necessary

If it's not too much of a hassle, I'd appreciate it but no worries if not

File: src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_omni/modeling_qwen2_5_omni.py:1590]: It means the input positions are not packed. Since we access `text_position_ids` below several times, we still need a variable set to smth

It has a huge comment block above explaining how packed multimodal positions look like, I will add a tiny remark here that second one is non-packed way :)

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py (hunk 2 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1312]: there was some failure in torch side or in compilation, can't remember now. Decided to simply fix and forget, as it won't hurt the performance

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1576]: yeah, kind of. This is the way positions are computed in `forward` pass and the reason why accuracy was low on benchmarks. It adds the past cached shift which is `+ cache_position[0]` and also multimodal shift which is `self.rope_deltas`

TBH this looks same as the computations before PR ("cache shifted positions + delta") but bench results are different, so we keep it identical to forward pass, including order of ops

File: src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py:830]: it has one new arg in input and modular can't handle it currently

File: tests/test_modeling_common.py (hunk 5 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4777]: will rebase after you merge on main

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1312]: Yea, no worries out of place usually is always safer anyways

File: src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py (hunk 4 above)
Comment (vasqu) [reply src/transformers/models/qwen2_5_vl/modular_qwen2_5_vl.py:830]: :(

File: tests/test_modeling_common.py (hunk 5 above)
Comment (vasqu) [reply tests/test_modeling_common.py:4777]: Should be merged on main

File: src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py (hunk 3 above)
Comment (pengzhenghao-nv) [reply src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py:1576]: no this is not the same.. `model_inputs["position_ids"]` provided by the super class seems to be wrong..
